lib/modules/practice_utils.rb

```diff
@@ -146,51 +146,49 @@ def fetch_bookmarks_by_practice(practice_id, duration = "30")
-        sql += " and dhs.status= $3"
-        records_array = ActiveRecord::Base.connection.exec_query(sql, "SQL", [[nil, "#{practice_id}"], [nil, "#{Time.now - duration.to_i.days}"], [nil, "#{status}"]])
-      else
-        records_array = ActiveRecord::Base.connection.exec_query(sql, "SQL", [[nil, "#{practice_id}"], [nil, "#{Time.now - duration.to_i.days}"]])
-      end
-    else
-      if(status.length > 0)
-        sql += " and dhs.status= $2"
-        records_array = ActiveRecord::Base.connection.exec_query(sql, "SQL", [[nil, "#{practice_id}"], [nil, "#{status}"]])
-      else
-        records_array = ActiveRecord::Base.connection.exec_query(sql, "SQL", [[nil, "#{practice_id}"]])
-      end
-    end
-    records_array[0]["count"]
+  def fetch_adoptions_by_practice_last_30_days(practice)
+    DiffusionHistory.get_with_practice(practice).where(created_at: Time.now.beginning_of_day - 30.days..Time.now)
   end
 
-  def fetch_adoptions_by_practice(practice_id, duration = "30")
-    sql = "select count(*) as the_count from diffusion_histories where practice_id = $1"
-    if duration == "30"
-      sql += " and created_at >= $2"
-      records_array = ActiveRecord::Base.connection.exec_query(sql, "SQL", [[nil, "#{practice_id}"], [nil, "#{Time.now - duration.to_i.days}"]])
-    else
-      records_array = ActiveRecord::Base.connection.exec_query(sql, "SQL", [[nil, "#{practice_id}"]])
-    end
-    records_array[0]["the_count"]
+  def fetch_adoptions_by_practice_all_time(practice)
+    DiffusionHistory.get_with_practice(practice)
   end
 
-  def fetch_adoption_facilities_for_practice(practice_id, duration="30", facility_data)
-    facility_ids = []
-    sql = "select va_facility_id from diffusion_histories where practice_id = $1"
-    if duration == "30"
-      sql += " and created_at >= $2"
-      records_array = ActiveRecord::Base.connection.exec_query(sql, "SQL", [[nil, "#{practice_id}"], [nil, "#{Time.now - duration.to_i.days}"]])
-    else
-      records_array = ActiveRecord::Base.connection.exec_query(sql, "SQL", [[nil, "#{practice_id}"]])
-    end
-    records_array.each do |rec|
-      facility_ids << facility_data.find(rec.values.first).station_number
+  def fetch_adoption_counts_by_status(status, adoptions)
+    case status
+    when 'Completed'
+      adoptions.select { |a| a_status = a.diffusion_history_statuses.first.status; a_status === 'Completed' || a_status === 'Implemented' || a_status === 'Complete' }.size
+    when 'In progress'
+      adoptions.select { |a| a_status = a.diffusion_history_statuses.first.status; a_status === 'In progress' || a_status === 'Planning' || a_status === 'Implementing' }.size
+    when 'Unsuccessful'
+      adoptions.select { |a| a_status = a.diffusion_history_statuses.first.status; a_status === 'Unsuccessful' }.size
     end
-    facility_ids
+  end
+
+  def fetch_adoptions_total_by_practice_and_status_last_30_days(practice, status)
+    adoptions = fetch_adoptions_by_practice_last_30_days(practice)
```

Comment:
I believe we can do `fetch_adoptions_by_practice_last_30_days(practice).get_by_successful_status.count` instead of using `select` above to get the count

Reply:
Ah yes, good catch! I should've utilized those scopes 😸  It has been refactored!